build failures with gcc 7.3.1

One CDash machine was upgraded to Fedora 27 and gcc 7.3.1 and it hit
several failures:

drheapstat-dbg-32 build is failing with gcc 7.3.1 upgrade:
http://dynamorio.org/CDash/buildSummary.php?buildid=35326
```
tests/operators.cpp line 111 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=35326)
/.../src/tests/operators.cpp:111:67: error: size of array is too large
tests/operators.cpp line 132 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=35326)
/.../src/tests/operators.cpp:132:81: error: size of array is too large
make[2]: *** [tests/CMakeFiles/operators.dir/build.make:63: tests/CMakeFiles/operators.dir/operators.cpp.o] Error 1
```

drmemory-dbg-32 is failing on:
http://dynamorio.org/CDash/buildSummary.php?buildid=35330
```
tests/memalign.c line 50 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=35330)
/.../src/tests/memalign.c:50:5: error: implicit declaration of function ‘cfree’; did you mean ‘free’? [-Werror=implicit-function-declaration]
tests/memalign.c line 60 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=35330)
/.../src/tests/memalign.c:60:9: error: implicit declaration of function ‘malloc_get_state’; did you mean ‘malloc_stats’? [-Werror=implicit-function-declatests/memalign.c line 60 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=35330)
/.../src/tests/memalign.c:60:7: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
make[2]: *** [tests/CMakeFiles/memalign.dir/build.make:63: tests/CMakeFiles/memalign.dir/memalign.c.o] Error 1
```

drmemory package is failing with:
http://dynamorio.org/CDash/buildSummary.php?buildid=35332
```
[ 17%] Built target dynamorio
Scanning dedrsyscall/linux_defines.h line 61 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=35332)
/.../src/drsyscall/linux_defines.h:61:11: fatal error: bits/siginfo.h: No such file or directory
drsyscall/linux_defines.h line 61 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=35332)
/.../src/drsyscall/linux_defines.h:61:11: fatal error: bits/siginfo.h: No such file or directory
drsyscall/linux_defines.h line 61 (http://dynamorio.org/CDash/viewBuildError.php?type=0&buildid=35332)
/.../src/drsyscall/linux_defines.h:61:11: fatal error: bits/siginfo.h: No such file or directory
/bin/ar: CMakeFiles/drsyscall_int.dir/drsyscall_linux.c.o: No such file or directory
```

--- a/drsyscall/linux_defines.h
+++ b/drsyscall/linux_defines.h
@@ -1,5 +1,5 @@
 /* **********************************************************
- * Copyright (c) 2011-2016 Google, Inc.  All rights reserved.
+ * Copyright (c) 2011-2018 Google, Inc.  All rights reserved.
  * Copyright (c) 2007-2010 VMware, Inc.  All rights reserved.
  * **********************************************************/
 
@@ -58,7 +58,7 @@
  */
 # define __need_siginfo_t
 # define __need_sigevent_t
-# include <bits/siginfo.h>
+# include <asm/siginfo.h>
 #else
 # include <signal.h>
 #endif

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -477,10 +477,17 @@ else ()
   set(cs2bug_res "cs2bugMTd")
   set(cs2bug_flags "/DSKIP_MISMATCH_DTR")
 endif ()
-newtest_ex(cs2bug cs2bug.cpp "" "" "" OFF ${cs2bug_res} 0)
+newtest_ex(cs2bug cs2bug.cpp "" "" ""  OFF ${cs2bug_res} 0)
 append_property_string(TARGET cs2bug COMPILE_FLAGS "${cs2bug_flags}")
 
-newtest_ex(operators operators.cpp "" "" "" OFF ""
+# On Linux this test is very slow trying to hit OOM.  We tweak the options to try
+# and speed it up.
+if (UNIX)
+  set(operators_dr_ops "-checklevel 0 -no_enable_reset -vm_size 64M")
+else ()
+  set(operators_dr_ops "")
+endif()
+newtest_ex(operators operators.cpp "" "" "${operators_dr_ops}" OFF ""
   # ignore exit code (b/c -replace_malloc calls dr_exit_process(1) in lieu of exception)
   "ANY")
 newtest(float float.c)

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -480,14 +480,16 @@ endif ()
 newtest_ex(cs2bug cs2bug.cpp "" "" ""  OFF ${cs2bug_res} 0)
 append_property_string(TARGET cs2bug COMPILE_FLAGS "${cs2bug_flags}")
 
-# On Linux this test is very slow trying to hit OOM.  We tweak the options to try
-# and speed it up.
+# On Linux this test is very slow trying to hit OOM and then leak checking.
+# We tweak the options to try and speed it up and not timeout on Travis.
 if (UNIX)
   set(operators_dr_ops "-checklevel 0 -no_enable_reset -vm_size 64M")
+  set(operators_drm_ops "-no_count_leaks")
 else ()
   set(operators_dr_ops "")
+  set(operators_drm_ops "")
 endif()
-newtest_ex(operators operators.cpp "" "" "${operators_dr_ops}" OFF ""
+newtest_ex(operators operators.cpp "" "${operators_drm_ops}" "${operators_dr_ops}" OFF ""
   # ignore exit code (b/c -replace_malloc calls dr_exit_process(1) in lieu of exception)
   "ANY")
 newtest(float float.c)
@@ -691,8 +693,9 @@ if (TOOL_DR_MEMORY)
         newtest_ex(wrap_cs2bug cs2bug.cpp "" "-no_replace_malloc" "" OFF ${cs2bug_res} 0)
       endif ()
     endif ()
-    newtest_nobuild_ex(wrap_operators operators "" "-no_replace_malloc"
-      "" OFF "operators" 0 "")
+    newtest_nobuild_ex(wrap_operators operators ""
+      "-no_replace_malloc;${operators_drm_ops}" "${operators_dr_ops}"
+      OFF "operators" 0 "")
   endif ()
 
   # shared by all suppress tests

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -1,5 +1,5 @@
 # **********************************************************
-# Copyright (c) 2010-2017 Google, Inc.  All rights reserved.
+# Copyright (c) 2010-2018 Google, Inc.  All rights reserved.
 # Copyright (c) 2009-2010 VMware, Inc.  All rights reserved.
 # **********************************************************
 
@@ -615,6 +615,10 @@ if (WIN32)
   endif (TOOL_DR_MEMORY)
 endif (WIN32)
 
+# The docs claim plain "-Wno-alloc-size-larger-than" should disable but it doesn't.
+CHECK_C_COMPILER_FLAG("-Wno-alloc-size-larger-than=4294967295"
+  HAVE_ALLOC_SIZE_WARNING_INTERNAL)
+
 if (TOOL_DR_MEMORY)
   if (NOT ARM) # XXX i#1726: port to ARM
     # PR 525807: test malloc stacks
@@ -686,8 +690,14 @@ if (TOOL_DR_MEMORY)
 
   # shared by all suppress tests
   tobuild(suppress suppress.c)
-  if (APPLE)
+  if (UNIX)
     append_test_compile_flags(suppress "-Wno-uninitialized")
+    if (HAVE_ALLOC_SIZE_WARNING_INTERNAL)
+      # XXX: I can't get gcc 7.3.0 to not warn on the CallocOverflow test no matter
+      # what -Wno-alloc-size-larger-than* I try.  The only solution I came up with is
+      # to live with warnings :(
+      append_test_compile_flags(suppress "-Wno-error")
+    endif ()
   endif ()
   if (WIN32)
     tobuild_lib(suppress-mod-foo suppress-mod-foo.c "" "/nodefaultlib /noentry")

--- a/tests/app_suite/CMakeLists.txt
+++ b/tests/app_suite/CMakeLists.txt
@@ -1,5 +1,5 @@
 # **********************************************************
-# Copyright (c) 2011-2016 Google, Inc.  All rights reserved.
+# Copyright (c) 2011-2018 Google, Inc.  All rights reserved.
 # **********************************************************
 
 # Dr. Memory: the memory debugger
@@ -46,6 +46,12 @@ add_executable(app_suite_tests
                "${GTEST_ROOT}/src/gtest_main.cc")
 
 if (UNIX)
+  if (HAVE_ALLOC_SIZE_WARNING_INTERNAL)
+    # XXX: I can't get gcc 7.3.1 to not warn on the CallocOverflow test no matter
+    # what -Wno-alloc-size-larger-than* I try.  The only solution I came up with is
+    # to live with warnings :(
+    append_compile_flags(app_suite_tests "-Wno-error")
+  endif ()
   if (APPLE)
     append_compile_flags(app_suite_tests "-DMACOS -DUNIX")
     # Avoid gtest-internal warnings

--- a/tests/memalign.c
+++ b/tests/memalign.c
@@ -1,5 +1,5 @@
 /* **********************************************************
- * Copyright (c) 2010-2015 Google, Inc.  All rights reserved.
+ * Copyright (c) 2010-2018 Google, Inc.  All rights reserved.
  * Copyright (c) 2010 VMware, Inc.  All rights reserved.
  * **********************************************************/
 
@@ -46,8 +46,10 @@ int main()
 #ifdef LINUX
     struct mallinfo info;
 
+# if 0 /* Removing since cfree was removed from glibc 2.26+ */
     p = malloc(37);
     cfree(p);
+# endif
 
     mallopt(M_MMAP_THRESHOLD, 32 * 1024);
 
@@ -57,8 +59,10 @@ int main()
 
     malloc_stats();
 
+# if 0 /* Removing since malloc_get_state was removed from glibc 2.25+ */
     p = malloc_get_state();
     free(p);
+# endif
 #elif defined(MACOS)
     /* Tests for malloc zones (i#1699) are in mac_zones.c */
 #endif

--- a/tests/memalign.pattern.res
+++ b/tests/memalign.pattern.res
@@ -1,5 +1,5 @@
 # **********************************************************
-# Copyright (c) 2015-2016 Google, Inc.  All rights reserved.
+# Copyright (c) 2015-2018 Google, Inc.  All rights reserved.
 # **********************************************************
 #
 # Dr. Memory: the memory debugger
@@ -19,30 +19,30 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 #
 Error #1: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:71
+memalign.c:75
 Note: refers to 0 byte(s) beyond last valid byte in prior malloc
 
 Error #2: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:87
+memalign.c:91
 Note: refers to 0 byte(s) beyond last valid byte in prior malloc
 
 Error #3: WARNING: heap allocation failed
-memalign.c:91
+memalign.c:95
 
 Error #4: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:100
+memalign.c:104
 
 %if !MACOS
 Error #5: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:108
+memalign.c:112
 Note: refers to 0 byte(s) beyond last valid byte in prior malloc
 %endif
 
 : UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:116
+memalign.c:120
 Note: refers to 0 byte(s) beyond last valid byte in prior malloc
 
 %if !MACOS
 : UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:125
+memalign.c:129
 %endif

--- a/tests/memalign.res
+++ b/tests/memalign.res
@@ -1,5 +1,5 @@
 # **********************************************************
-# Copyright (c) 2010-2016 Google, Inc.  All rights reserved.
+# Copyright (c) 2010-2018 Google, Inc.  All rights reserved.
 # Copyright (c) 2009-2010 VMware, Inc.  All rights reserved.
 # **********************************************************
 #
@@ -20,54 +20,54 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 #
 Error #1: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:70
+memalign.c:74
 Note: refers to 1 byte(s) before next malloc
 
 Error #2: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:71
+memalign.c:75
 Note: refers to 0 byte(s) beyond last valid byte in prior malloc
 
 Error #3: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:86
+memalign.c:90
 Note: refers to 1 byte(s) before next malloc
 
 Error #4: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:87
+memalign.c:91
 Note: refers to 0 byte(s) beyond last valid byte in prior malloc
 
 Error #5: WARNING: heap allocation failed
-memalign.c:91
+memalign.c:95
 
 Error #6: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:99
+memalign.c:103
 Note: refers to 1 byte(s) before next malloc
 
 Error #7: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:100
+memalign.c:104
 
 %if !MACOS
 Error #8: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:107
+memalign.c:111
 Note: refers to 1 byte(s) before next malloc
 
 Error #9: UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:108
+memalign.c:112
 Note: refers to 0 byte(s) beyond last valid byte in prior malloc
 %endif
 
 : UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:115
+memalign.c:119
 Note: refers to 1 byte(s) before next malloc
 
 : UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:116
+memalign.c:120
 Note: refers to 0 byte(s) beyond last valid byte in prior malloc
 
 %if !MACOS
 : UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:123
+memalign.c:127
 Note: refers to 1 byte(s) before next malloc
 
 : UNADDRESSABLE ACCESS beyond heap bounds: reading 1 byte(s)
-memalign.c:125
+memalign.c:129
 %endif

--- a/tests/operators.cpp
+++ b/tests/operators.cpp
@@ -1,5 +1,5 @@
 /* **********************************************************
- * Copyright (c) 2012-2017 Google, Inc.  All rights reserved.
+ * Copyright (c) 2012-2018 Google, Inc.  All rights reserved.
  * **********************************************************/
 
 /* Dr. Memory: the memory debugger
@@ -28,13 +28,9 @@
 #include "stdlib.h"
 #include "limits.h"
 
-// Windows compiler and gcc4.4.3 complain if the size is over 0x7fffffff
-#if defined(MACOS) || defined(__clang__)
-/* clang won't let it get bigger */
-# define SIZE_OOM 0xfffffff
-#else
-# define SIZE_OOM 0x7fffffff
-#endif
+// Windows compiler and gcc4.4.3 complain if the size is over 0x7fffffff;
+// Clang and gcc 7.3.2 complain if it's over 0xfffffff.
+#define SIZE_OOM 0xfffffff
 #define SIZE_OOM_ARR 1024
 #ifdef UNIX
 # define IF_UNIX_ELSE(x,y) x
@@ -129,7 +125,7 @@ test_throw()
         // On Linux if the size is too small we don't run out of memory until
         // we've constructed most of the elements, which takes a long
         // time and causes issues under drmem w/ DR doing resets, etc.
-        hasdtr *lots = new hasdtr[IF_UNIX_ELSE(SIZE_OOM,SIZE_OOM/sizeof(hasdtr))];
+        hasdtr *lots = new hasdtr[IF_UNIX_ELSE(SIZE_OOM/2,SIZE_OOM/sizeof(hasdtr))];
         lots[0].y = 4;
     } catch (std::bad_alloc&) {
         std::cout << "caught bad_alloc" << std::endl;

--- a/tests/operators.cpp
+++ b/tests/operators.cpp
@@ -30,7 +30,11 @@
 
 // Windows compiler and gcc4.4.3 complain if the size is over 0x7fffffff;
 // Clang and gcc 7.3.2 complain if it's over 0xfffffff.
-#define SIZE_OOM 0xfffffff
+#ifdef UNIX
+# define SIZE_OOM 0xfffffff
+#else
+# define SIZE_OOM 0x7fffffff
+#endif
 #define SIZE_OOM_ARR 1024
 #ifdef UNIX
 # define IF_UNIX_ELSE(x,y) x
@@ -111,7 +115,7 @@ test_nothrow()
         }
     }
     for (unsigned int j = 0; j <= i && j < SIZE_OOM_ARR; j++)
-        delete p[j];
+        delete[] p[j];
     delete[] p;
 }
 

--- a/tests/operators.cpp
+++ b/tests/operators.cpp
@@ -136,7 +136,8 @@ test_throw()
     }
 }
 
-int main()
+int
+main()
 {
     test_placement();
 

--- a/tests/operators.cpp
+++ b/tests/operators.cpp
@@ -41,6 +41,11 @@
 #else
 # define IF_UNIX_ELSE(x,y) y
 #endif
+#ifdef X64
+# define IF_X64_ELSE(x,y) x
+#else
+# define IF_X64_ELSE(x,y) y
+#endif
 
 class hasdtr {
 public:
@@ -129,7 +134,10 @@ test_throw()
         // On Linux if the size is too small we don't run out of memory until
         // we've constructed most of the elements, which takes a long
         // time and causes issues under drmem w/ DR doing resets, etc.
-        hasdtr *lots = new hasdtr[IF_UNIX_ELSE(SIZE_OOM/2,SIZE_OOM/sizeof(hasdtr))];
+        // gcc 7.3+ forces us to use the new smaller SIZE_OOM/2 for 32-bit,
+        // but that causes 64-bit to take forever here so we up it.
+        hasdtr *lots = new hasdtr[IF_UNIX_ELSE(IF_X64_ELSE(0x7fffffff,SIZE_OOM/2),
+                                               SIZE_OOM/sizeof(hasdtr))];
         lots[0].y = 4;
     } catch (std::bad_alloc&) {
         std::cout << "caught bad_alloc" << std::endl;

--- a/tests/operators.out
+++ b/tests/operators.out
@@ -1,5 +1,5 @@
 # **********************************************************
-# Copyright (c) 2012-2017 Google, Inc.  All rights reserved.
+# Copyright (c) 2012-2018 Google, Inc.  All rights reserved.
 # **********************************************************
 #
 # Dr. Memory: the memory debugger
@@ -30,5 +30,7 @@ about to die
 %if X64
 ~~Dr.M~~       1 unique,     1 total warning(s)
 %endif
+%if WINDOWS
 ~~Dr.M~~       0 unique,     0 total,      0 byte(s) of leak(s)
 ~~Dr.M~~       0 unique,     0 total,      0 byte(s) of possible leak(s)
+%endif

--- a/tests/operators.res
+++ b/tests/operators.res
@@ -20,9 +20,9 @@
 #
 %if X32
 : WARNING: heap allocation failed
-operators.cpp:97
+operators.cpp:102
 : WARNING: heap allocation failed
-operators.cpp:111
+operators.cpp:116
 %endif
 : WARNING: heap allocation failed
-operators.cpp:132
+operators.cpp:140

